### PR TITLE
PEP 659: Specializing Adaptive Interpreter

### PR DESCRIPTION
**Pull request by markshannon**

### CODE REVIEW

File: pep-0659.rst [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+Adding a specializing, adaptive interpreter to CPython will bring significant performance improvements.
+It is hard to come up with meaningful numbers, as it depends very much on the benchmarks and on work that has not yet happened.
+Extensive experimentation suggests speedups of up to 50%.
+Even if the speedup were only 25%, this would still be a worthwhile enhancement.
+
+Motivation
+==========
+
+Python is widely acknowledged as slow.
+Whilst Python will never attain the performance of low-level langauges like C, Fortran, or even Java,
```
Comment (JelleZijlstra) [inline pep-0659.rst:36]:
```suggestion
Whilst Python will never attain the performance of low-level languages like C, Fortran, or even Java,
```

File: pep-0659.rst [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+However, specialization is the most important, both in itself and as an enabler of other optimizations.
+Therefore it makes sense to focus our efforts on specialization first, if we want to improve the performance of CPython.
+
+Specialization is typically done in the context of a JIT compiler, but research shows specialization in an interpreter
+can boost performance significantly, even outperforming a naive compiler [1]_.
+
+There have been several ways of doing this proposed in the academic literature,
+but most attempt to optimize regions larger than a single bytecode [1]_ [2]_.
+Using larger regions than a single instruction, requires code to handle deoptimization in the middle of a region.
+Specialization at the level on individual bytecodes makes deoptimization trivial, as it cannot occur in the middle of a region.
```
Comment (JelleZijlstra) [inline pep-0659.rst:67]:
```suggestion
Specialization at the level of individual bytecodes makes deoptimization trivial, as it cannot occur in the middle of a region.
```

File: pep-0659.rst [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+
+By speculatively specializing individual bytecodes, we can gain significant performance improvements without anything but the most local,
+and trivial to implement, de-optimizations.
```
Comment (JelleZijlstra) [inline pep-0659.rst:70]:
```suggestion
and trivial to implement, deoptimizations.
```

File: pep-0659.rst [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+Any instruction that would benefit from specialization will be replaced by an "adaptive" form of that instruction.
+When executed, the adaptive instructions will specialize themselves in response to the types and values that they see.
+
+Quickening
+----------
+
+Quickening is the process of replacing slow instructions with faster variants.
+
+Quickened code has number of advantages over the normal bytecode:
```
Comment (JelleZijlstra) [inline pep-0659.rst:105]:
```suggestion
Quickened code has a number of advantages over the normal bytecode:
```

File: pep-0659.rst [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+
+* It can be changed at runtime
+* It can use super-instructions that span lines and take multiple operands.
+* It does not need to handle tracing as it can fallback to the normal bytecode for that.
+
+In order that tracing can supported, and quickening performed quickly, the quickened instruction format should match the normal
```
Comment (JelleZijlstra) [inline pep-0659.rst:111]:
```suggestion
In order that tracing can be supported, and quickening performed quickly, the quickened instruction format should match the normal
```

File: pep-0659.rst [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+If the counter reaches the minimum value, the instruction is de-optimized by simply replacing its opcode with the adaptive version.
+
+Ancillary data
+--------------
+
+Most families of specialized instructions will require more information than can fit in an 8-bit operand.
+To do this, an array of specialization data entries will be maintained alongside the new instruction array.
+For instructions that need specialization data, the operand in the quickened array will serve as a partial index, 
+along with the offset of the instruction, to find the first specialization data entry for that instruction.
+Each entry will be 8 bytes (for a 64 bit machine). The data in a entry, and the number of entries needed, will vary from instruction to instruction.
```
Comment (JelleZijlstra) [inline pep-0659.rst:144]:
```suggestion
Each entry will be 8 bytes (for a 64 bit machine). The data in an entry, and the number of entries needed, will vary from instruction to instruction.
```

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
Added some wording suggestions.

**Review by WM7586 — APPROVED** [review]

**Review by WM7586 — CHANGES_REQUESTED** [review]
pep-0659.rst

**Comment by CAM-Gerlach** [comment]
@python/organization-owners Spam x2 :point_up_2:

**Pull request merged**